middleware/base_rate_limiter.rb

```diff
@@ -0,0 +1,152 @@
+          @data[user_guid] = request_count
+        end
+      end
+
+      private
+
+      def create_new_request_count(user_guid, reset_interval_in_minutes)
+        requests = 0
+        valid_until = next_reset_interval(user_guid, reset_interval_in_minutes)
+        @data[user_guid] = RequestCount.new(requests, valid_until)
+        [requests, valid_until]
+      end
+    end
+
+    class RateLimitHeaders
+      attr_accessor :limit, :reset, :remaining
+
+      def initialize(suffix)
+        @prefix = 'X-RateLimit'
+        @suffix = suffix.nil? ? nil : '-' + suffix
+        @limit = nil
+        @reset = nil
+        @remaining = nil
+      end
+
+      def as_hash
+        return {} if [@limit, @reset, @remaining].all?(&:nil?)
+
+        { "#{@prefix}-Limit#{@suffix}" => @limit, "#{@prefix}-Reset#{@suffix}" => @reset, "#{@prefix}-Remaining#{@suffix}" => @remaining }
+      end
+    end
+
+    class BaseRateLimiter
+      include CloudFoundry::Middleware::ClientIp
+
+      def initialize(app, logger, request_counter, reset_interval, header_suffix=nil)
+        @app = app
+        @logger = logger
+        @request_counter = request_counter
+        @reset_interval = reset_interval
+        @header_suffix = header_suffix
+      end
+
+      def call(env)
+        rate_limit_headers = RateLimitHeaders.new(@header_suffix)
+
+        request = ActionDispatch::Request.new(env)
+
+        unless skip_rate_limiting?(env, request)
+          user_guid = user_token?(env) ? env['cf.user_guid'] : client_ip(request)
+
+          count, valid_until = @request_counter.get(user_guid, @reset_interval, @logger)
+          new_request_count = count + 1
+
+          rate_limit_headers.limit = global_request_limit(env).to_s
+          rate_limit_headers.reset = valid_until.to_i.to_s
+          rate_limit_headers.remaining = estimate_remaining(env, new_request_count)
+
+          return too_many_requests!(env, rate_limit_headers) if exceeded_rate_limit(new_request_count, env)
+
+          @request_counter.increment(user_guid)
+        end
+
+        status, headers, body = @app.call(env)
+        [status, headers.merge(rate_limit_headers.as_hash), body]
+      end
+
+      private
+
+      def skip_rate_limiting?(env, request)
+        raise NotImplementedError
+      end
+
+      def global_request_limit(env)
+        raise NotImplementedError
+      end
+
+      def rate_limit_error(env)
+        raise NotImplementedError
+      end
+
+      def exceeded_rate_limit(count, env)
+        count > per_process_request_limit(env)
+      end
+
+      def estimate_remaining(env, new_count)
+        global_limit = global_request_limit(env)
+        limit = per_process_request_limit(env)
+        return '0' unless limit > 0
+
+        estimate = ((limit - new_count).to_f / limit).floor(1) * global_limit
+        [0, estimate].max.to_i.to_s
+      end
+
+      def user_token?(env)
+        !!env['cf.user_guid']
+      end
+
+      def basic_auth?(auth)
```

Comment:
We can change the signature to `def basic_auth?(env)` and put the following code into this method, as it's not used elsewhere:
```
auth = Rack::Auth::Basic::Request.new(env)
```